File: app/src/main/java/com/ortin/gesturetranslator/di/FeatureModule.java
package com.ortin.gesturetranslator.di;

import android.content.Context;

import com.ortin.gesturetranslator.domain.repository.LoadImageRepository;
import com.ortin.gesturetranslator.domain.repository.WordCompilerRepository;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.components.SingletonComponent;

@Module
@InstallIn(SingletonComponent.class)
public class FeatureModule {
    @Provides
    @Singleton
    LoadImageRepository provideLoadImageRepository(CameraManager cameraManager) {
        return new LoadImageRepositoryImpl(cameraManager);
    }

    @Provides
    @Singleton
    CameraManager provideCameraManager(@ApplicationContext Context context) {
        return new CameraManagerImpl(context);
    }

    @Provides
    @Singleton
    WordCompilerRepository provideWordCompilerRepository(WordCompilerManager wordCompilerManager) {
        return new WordCompilerRepositoryImpl(wordCompilerManager);
    }

    @Provides
    @Singleton
    WordCompilerManager provideWordCompilerManager() {
        return new WordCompilerManagerImpl();
    }
}

File: app/src/main/kotlin/com/ortin/gesturetranslator/app/di/FeatureModule.kt
package com.ortin.gesturetranslator.app.di

import com.ortin.gesturetranslator.domain.repository.LoadImageRepository
import com.ortin.gesturetranslator.domain.repository.WordCompilerRepository
import com.ortin.gesturetranslator.feature.managers.camera.CameraManager
import com.ortin.gesturetranslator.feature.managers.camera.CameraManagerImpl
import com.ortin.gesturetranslator.feature.managers.word_compiler.WordCompilerManager
import com.ortin.gesturetranslator.feature.managers.word_compiler.WordCompilerManagerImpl
import com.ortin.gesturetranslator.feature.repository.LoadImageRepositoryImpl
import com.ortin.gesturetranslator.feature.repository.WordCompilerRepositoryImpl
import dagger.Binds
import dagger.Module
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
interface FeatureModuleBind {
    @Binds
    @Singleton
    fun bindCameraManager(impl: CameraManagerImpl): CameraManager

    @Binds
    @Singleton
    fun bindLoadImageRepository(impl: LoadImageRepositoryImpl): LoadImageRepository

    @Binds
    @Singleton
    fun bindWordCompilerManager(impl: WordCompilerManagerImpl): WordCompilerManager

    @Binds
    @Singleton
    fun bindWordCompilerRepository(impl: WordCompilerRepositoryImpl): WordCompilerRepository
}
